### ISSUE TITLE
Feature: deprecate an option/command

### ISSUE THREAD

**Issue by laggingreflex** [yargs/yargs#1518]
Allow setting `deprecated: true` for an option or a command, like:

```
yargs.options({
  oldOption: {type: 'boolean', deprecated: true }
  newOption: {type: 'boolean' }
})
```

which will show up in help as:

```
Options:
  --old-option [deprecated]
  --new-option
```

**Comment by bcoe** [comment]
@laggingreflex this is a neat idea 👍

**Comment by laggingreflex** [comment]
Cool. I'll try to implement myself this when I get some time. 

Any preferences for the following: 

* Key - "deprecated" 

* Type - I'm thinking boolean "true/false", but also a string like "Please use --newOption instead"

* Placing of the text - 

  Boolean

  1. `[deprecated] --option description [type] [default]`

  1. `--option [deprecated] description [type] [default]`

  1. `--option description [deprecated] [type] [default]`

  String

  1. `... [deprecated: text] ...`

  1. `... [deprecated - text] ...`

* Should `[type]` and `[default]` be hidden when `deprecated` is set? I was thinking they should be hidden as they're not gonna be needed anyway, but maybe that decision should be left to the user (they can always remove those fields themselves).

* Should all the deprecated options be pushed below the normal options? I'd prefer this. 

Anything else?

**Comment by mleguen** [comment]
@laggingreflex Thanks for digging into this.

`deprecated` sounds good to me.

A boolean would be easier to implement, but a string would be more practical (nowing how not to use a deprecated option is good, and IMHO the first reason for having such a deprecation notice).

Perhaps you should have a look at existing keys, to try to mimic as much as possible the behavior and naming of an existing key, for homogeneity (I do not have the time right now to do it myself)?

Concerning text placement, I would suggest looking first as how "reference" CLI (GNU tools, git, docker, npm, etc.) do it (if they do it), not to reinvent the wheel.

I would also suggest keeping type and default, as a deprecated option is still valid and usable.

And if we add this, I would definitely want it also for commands!

**Comment by laggingreflex** [comment]
Was able to finally get this into motion. Please checkout #1559

It implements deprecating just the option for now. Will try to work on command later. 

Also, I couldn't find much references in common tools (git, docker, npm) that do it. If you come across it please let me know. 

I did manage to find a deprecated option in Docker, but it actually didn't mention that option at all in its --help. 

However running the command using the deprecated option logged out a warning saying "this option has been deprecated". Is that something that should be implemented? Maybe in addition to what I've already done above?

**Comment by shadowspawn** [comment]
Looks like this landed in #1624, thanks @laggingreflex.

This issue has not had any activity in over six months.

Feel free to open a new issue if it comes up again, with new information and renewed interest.
 
Thank you  for your contributions.
